### PR TITLE
[Abort] Add convrnnt.py

### PR DESCRIPTION
**Pull request by yfyeung**
Add ConvRNN-T Encoder
ConvRNN-T: Convolutional Augmented Recurrent Neural Network Transducers for Streaming Speech Recognition
https://arxiv.org/pdf/2209.14868.pdf

model size: 44M

The best WER on LibriSpeech 960h within 20 epoch is:
epoch-20 avg-4 modified_beam_search beam-size-4 use-averaged-model
-- | test-clean | test-other
-- | -- | --
WER | 5.01 | 11.92

### CODE REVIEW

File: egs/librispeech/ASR/conv_transducer_stateless/convrnnt.py [hunk 1]
```diff
@@ -0,0 +1,958 @@
...
+from encoder_interface import EncoderInterface
+from scaling import (
+    ActivationBalancer,
+    BasicNorm,
+    DoubleSwish,
+    ScaledConv1d,
+    ScaledConv2d,
+    ScaledLinear,
+)
+from torch import Tensor, nn
```
Comment (marcoyang1998) [inline egs/librispeech/ASR/conv_transducer_stateless/convrnnt.py:15]: I suggest using `torch.Tensor` for consistency. Currently, you are using both `Tensor` and `torch.Tensor` for return types.

Comment (yfyeung) [reply egs/librispeech/ASR/conv_transducer_stateless/convrnnt.py:15]: OK, I will modify it.

File: egs/librispeech/ASR/conv_transducer_stateless/convrnnt.py [hunk 2]
```diff
@@ -0,0 +1,1214 @@
...
+        # Conv5
+        # (N, 64, C, T) -> (N, 1, C, T) -> (N, C, T)
+        x = self.conv5(x)
+        x = x.squeeze(1).permute(2, 0, 1)
+        return x
+
+
+class SEModule(nn.Module):
+    """SE Module as defined in original SE-Nets with a few additions
+    Modified from rwightman`s squeeze_excite.py
```
Comment (csukuangfj) [inline egs/librispeech/ASR/conv_transducer_stateless/convrnnt.py:510]: please note down the link to the original implementation.

File: egs/librispeech/ASR/conv_transducer_stateless/convrnnt.py [hunk 3]
```diff
@@ -0,0 +1,1214 @@
...
+        rd_channels=None,
+        rd_divisor=8,
+        bias=True,
+    ):
+        super(SEModule, self).__init__()
+        if not rd_channels:
+            rd_channels = self.make_divisible(
+                channels * rd_ratio, rd_divisor, round_limit=0.0
+            )
+        self.fc1 = nn.Conv1d(channels, rd_channels, kernel_size=1, bias=bias)
```
Comment (csukuangfj) [inline egs/librispeech/ASR/conv_transducer_stateless/convrnnt.py:526]: Why not use `nn.Linear` to replace `nn.Conv1d`?

File: egs/librispeech/ASR/conv_transducer_stateless/convrnnt.py [hunk 4]
```diff
@@ -0,0 +1,1214 @@
...
+    """
+
+    def __init__(
+        self,
+        channels,
+        rd_ratio=1.0 / 16,
+        rd_channels=None,
+        rd_divisor=8,
+        bias=True,
+    ):
```
Comment (csukuangfj) [inline egs/librispeech/ASR/conv_transducer_stateless/convrnnt.py:520]: Please document the meaning of each argument.

File: egs/librispeech/ASR/conv_transducer_stateless/convrnnt.py [hunk 5]
```diff
@@ -0,0 +1,1214 @@
...
+        if not rd_channels:
+            rd_channels = self.make_divisible(
+                channels * rd_ratio, rd_divisor, round_limit=0.0
+            )
+        self.fc1 = nn.Conv1d(channels, rd_channels, kernel_size=1, bias=bias)
+        self.act = nn.ReLU()
+        self.fc2 = nn.Conv1d(rd_channels, channels, kernel_size=1, bias=bias)
+        self.gate = nn.Sigmoid()
+
+    def forward(self, x, src_key_padding_mask):
```
Comment (csukuangfj) [inline egs/librispeech/ASR/conv_transducer_stateless/convrnnt.py:531]: What is `x` and `src_key_padding_mask`?
Please add documentation.

File: egs/librispeech/ASR/conv_transducer_stateless/convrnnt.py [hunk 6]
```diff
@@ -0,0 +1,1214 @@
...
+        x_se_num = src_key_padding_mask.eq(False).sum(axis=1, keepdim=True)
+        x_se_num = x_se_num.unsqueeze(1)
+        x_se = x.sum(axis=2, keepdim=True) / x_se_num
```
Comment (csukuangfj) [inline egs/librispeech/ASR/conv_transducer_stateless/convrnnt.py:534]: Do you need to remove the paddings when computing the `sum`?

File: egs/librispeech/ASR/conv_transducer_stateless/convrnnt.py [hunk 7]
```diff
@@ -0,0 +1,1214 @@
...
+        channels,
+        rd_ratio=1.0 / 16,
+        rd_channels=None,
+        rd_divisor=8,
+        bias=True,
+    ):
+        super(SEModule, self).__init__()
+        if not rd_channels:
+            rd_channels = self.make_divisible(
+                channels * rd_ratio, rd_divisor, round_limit=0.0
```
Comment (csukuangfj) [inline egs/librispeech/ASR/conv_transducer_stateless/convrnnt.py:524]: This class can be simplified a lot by constraining that
```
channels % rd_channels == 0
```

**Review by csukuangfj — CHANGES_REQUESTED** [review]

File: egs/librispeech/ASR/conv_transducer_stateless/convrnnt.py (hunk 2 above)
Comment (yfyeung) [reply egs/librispeech/ASR/conv_transducer_stateless/convrnnt.py]: ok, I will fix it.

File: egs/librispeech/ASR/conv_transducer_stateless/convrnnt.py (hunk 3 above)
Comment (yfyeung) [reply egs/librispeech/ASR/conv_transducer_stateless/convrnnt.py]: ok, I will fix it.

File: egs/librispeech/ASR/conv_transducer_stateless/convrnnt.py (hunk 4 above)
Comment (yfyeung) [reply egs/librispeech/ASR/conv_transducer_stateless/convrnnt.py:517]: ok, I will fix it.

File: egs/librispeech/ASR/conv_transducer_stateless/convrnnt.py (hunk 5 above)
Comment (yfyeung) [reply egs/librispeech/ASR/conv_transducer_stateless/convrnnt.py:526]: ok, I will fix it.

File: egs/librispeech/ASR/conv_transducer_stateless/convrnnt.py (hunk 6 above)
Comment (yfyeung) [reply egs/librispeech/ASR/conv_transducer_stateless/convrnnt.py]: ok, I will fix it.

File: egs/librispeech/ASR/conv_transducer_stateless/convrnnt.py (hunk 7 above)
Comment (yfyeung) [reply egs/librispeech/ASR/conv_transducer_stateless/convrnnt.py]: ok, I will fix it.

**Comment by yfyeung** [comment]
**ConvRNN-T Model**
The WER using modified beam search with beam size 4 is:
epoch-12-avg-2
test-clean | test-other
-- | --
WER | 5.44 | 12.73

**Comment by yfyeung** [comment]
**ConvRNN-T Model**
The WER using modified beam search with beam size 4 is:
epoch-12-avg-2
-- | test-clean | test-other
-- | -- | --
WER | 5.44 | 12.73

**Comment by yfyeung** [comment]
**ConvRNN-T Model**
44M, streaming model
The WER of epoch-12-avg-2 using modified beam search with beam size 4 is:
test-clean | test-other
-- | --
WER | 5.44 | 12.73

**Comment by yfyeung** [comment]
**ConvRNN-T Model**
44M, streaming model
The WER of epoch-12-avg-2 (not best) using modified beam search with beam size 4 is

-- | test-clean | test-other
-- | -- | --
WER | 5.44 | 12.73

**Comment by wangtiance** [comment]
Model Clean Other Size (M)
RNN-T 5.9 15.71 30
Conformer 5.7 14.24 29
ContextNet 6.02 14.42 28
ConvRNN-T 5.11 13.82 29

The WER shown in the paper seems a lot worse than the original papers of conformer/contextnet etc. Any idea why is that?

**Comment by yfyeung** [comment]
> Model Clean Other Size (M) RNN-T 5.9 15.71 30 Conformer 5.7 14.24 29 ContextNet 6.02 14.42 28 ConvRNN-T 5.11 13.82 29
> 
> The WER shown in the paper seems a lot worse than the original papers of conformer/contextnet etc. Any idea why is that?

I can't reproduce Google's setup. So, I have no no idea.

File: egs/librispeech/ASR/conv_transducer_stateless/train.py [hunk 8]
```diff
@@ -0,0 +1,1146 @@
...
+    checkpoints = load_checkpoint_if_available(
+        params=params, model=model, model_avg=model_avg
+    )
+
+    model.to(device)
+    if world_size > 1:
+        logging.info("Using DDP")
+        model = DDP(model, device_ids=[rank])
+
+    optimizer = Eve(model.parameters(), lr=params.initial_lr)
```
Comment (danpovey) [inline egs/librispeech/ASR/conv_transducer_stateless/train.py:967]: Probably ScaledAdam would be a bit better than this.

**Pull request closed without merging**